source/h2_frames.c

```diff
@@ -0,0 +1,1618 @@
+        return AWS_OP_ERR;
+    }
+    /* Write flags */
+    if (!aws_byte_buf_write_u8(output, flags)) {
+        return AWS_OP_ERR;
+    }
+    /* Write stream id (with reserved first bit) */
+    if (!aws_byte_buf_write_be32(output, header->stream_id & s_31_bit_mask)) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/***********************************************************************************************************************
+ * Encoder
+ **********************************************************************************************************************/
+int aws_h2_frame_encoder_init(struct aws_h2_frame_encoder *encoder, struct aws_allocator *allocator) {
+    AWS_PRECONDITION(encoder);
+    AWS_PRECONDITION(allocator);
+
+    AWS_ZERO_STRUCT(*encoder);
+    encoder->allocator = allocator;
+
+    encoder->hpack = aws_hpack_context_new(allocator, s_hpack_dynamic_table_initial_size);
+    if (!encoder->hpack) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
+    AWS_PRECONDITION(encoder);
+
+    aws_hpack_context_destroy(encoder->hpack);
+}
+
+/***********************************************************************************************************************
+ * Decoder
+ **********************************************************************************************************************/
+int aws_h2_frame_decoder_init(struct aws_h2_frame_decoder *decoder, struct aws_allocator *allocator) {
+    AWS_PRECONDITION(decoder);
+    AWS_PRECONDITION(allocator);
+
+    AWS_ZERO_STRUCT(*decoder);
+    decoder->allocator = allocator;
+
+    decoder->hpack = aws_hpack_context_new(allocator, s_hpack_dynamic_table_initial_size);
+    if (!decoder->hpack) {
+        goto failed_create_hpack;
+    }
+
+    if (aws_byte_buf_init(&decoder->header_scratch, allocator, 512)) {
+        goto failed_init_header_scratch;
+    }
+
+    return AWS_OP_SUCCESS;
+
+failed_init_header_scratch:
+    aws_hpack_context_destroy(decoder->hpack);
+
+failed_create_hpack:
+    return AWS_OP_ERR;
+}
+void aws_h2_frame_decoder_clean_up(struct aws_h2_frame_decoder *decoder) {
+    AWS_PRECONDITION(decoder);
+
+    aws_byte_buf_clean_up(&decoder->header_scratch);
+
+    aws_hpack_context_destroy(decoder->hpack);
+}
+
+int aws_h2_frame_decoder_begin(struct aws_h2_frame_decoder *decoder, struct aws_byte_cursor *data) {
+
+    AWS_PRECONDITION(decoder);
+    AWS_PRECONDITION(data);
+
+    const struct aws_byte_cursor data_init = *data;
+
+    /* Reset the scratch buffer */
+    decoder->header_scratch.len = 0;
+
+    /* Read length */
+    uint32_t payload_len = 0;
+    uint8_t *length_ptr = ((uint8_t *)&payload_len);
+    if (!aws_byte_cursor_read(data, length_ptr, 3)) {
+        goto read_error;
+    }
+    payload_len = aws_ntoh24(payload_len);
+    /* Assert top byte isn't set */
+    AWS_FATAL_ASSERT((payload_len & 0xFF000000) == 0);
```

Comment:
Long term, should we just return failure here? Otherwise, someone could DOS a service by feeding bad payload lengths

Reply:
Nah, I'm only reading 3 bytes from the stream (see line 471), so this is really just a sanity check.